### ISSUE TITLE
RT-DETR train error

### ISSUE THREAD

**Issue by whatisslove11** [ultralytics/ultralytics#5978]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

@glenn-jocher hello!
I wanna train RT-DETR model , but I have this error:
```
Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
/usr/local/lib/python3.10/dist-packages/torch/autograd/__init__.py:200: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:71.)
```

When I trained any YOLO model, I didn't get this error
I left my training code below

### Minimal Reproducible Example

```
import torch
from ultralytics import RTDETR, YOLO
import wandb

import random
import numpy as np
import os

def seed_everything(seed):
    random.seed(seed)
    os.environ['PYTHONHASHSEED'] = str(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)

seed_everything(seed=42)

args = dict(
   epochs=200,
   batch=64,
   data=f'/home/jupyter/datasphere/datasets/guns_next_dataset/guns_hacaton_aug/data.yaml',
   name=rtdetr-x',
   seed=42,
   device='0'
)

model = YOLO('rtdetr-x.pt')
model.train(**args)
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by RizwanMunawar** [comment]
@whatisslove11, we appreciate you sharing this information. We will conduct internal testing and promptly provide you with updates. Thanks

**Comment by gurkirt** [comment]
I am receiving the same warning!

**Comment by glenn-jocher** [comment]
@gurkirt thanks for reaching out and sorry for the inconvenience. This warning is related to the backward pass in CUDA during training. Specifically, `grid_sampler_2d_backward_cuda` does not have a deterministic implementation, and when you have set 'torch.use_deterministic_algorithms' to True, this warning is raised.

Although this is a warning and not an error, it means that a certain operation cannot guarantee the same output for multiple runs even when using the same input and initial states, which might interfere with reproducibility. This is typically due to non-deterministic algorithms used in CUDA operations that have a multi-threaded nature.

You can try setting 'torch.use_deterministic_algorithms' to False in your code, but this may impact the reproducibility of your results.

We're aware of this and currently looking for a workaround or fix. We appreciate your patience as we improve our software. Thank you!

**Comment by whatisslove11** [comment]
@glenn-jocher hello! Unfortunately, I tried to set torch.use_deterministic_algorithms = False immediately after the warning appeared, and it didn't work for me. The warning continued to appear further

**Comment by glenn-jocher** [comment]
@whatisslove11, apologies for the confusion, and I appreciate your patience.

Setting `torch.use_deterministic_algorithms = False` would not solve the problem once the training has started and the warning thrown. This setting needs to be done before starting the training process.

Please implement it at the start of your script, after importing the required libraries. However, please note that as a consequence of this setting, certain operations might behave in a non-deterministic manner which might slightly vary the outcomes across different training runs, even with the same initial conditions.

Remember, this is a warning and not an error, so it shouldn't halt your training process. As I mentioned earlier, we are aware of this issue and actively working towards a solution. I will update you as soon as we make progress.

Thank you again for your understanding and cooperation.

**Comment by glenn-jocher** [comment]
@whatisslove11 @RizwanMunawar after reviewing the above I realized you are using the wrong class. To train RTDETR models you use the RTDETR class, not the YOLO class:

```python
from ultralytics import RTDETR

model = RTDETR()
...
```

See https://docs.ultralytics.com/models/rtdetr/ for details

<img width="882" alt="Screenshot 2023-11-06 at 14 00 12" src="https://github.com/ultralytics/ultralytics/assets/26833433/ae948b43-7301-4aa9-9567-711743cfff81">

**Comment by whatisslove11** [comment]
@glenn-jocher hello!
I decided to train RT-DETR with your correction, but this warning didn't disappear
Do I need to provide notebook/minimal code again?

**Comment by glenn-jocher** [comment]
@whatisslove11, thank you for your due diligence in applying the correction. The warning you are experiencing is due to the deterministic behavior enforcement in PyTorch while a non-deterministic CUDA operation is being used. Changing the model class to RTDETR is correct for training RT-DETR models, but it will not affect this warning.

To clarify, the warning is informational and meant to alert you that reproducibility might be affected due to the non-determinism of certain GPU operations. It should not prevent the training from proceeding. Even with this warning, in most cases, the model will train successfully and you can continue your work without any immediate issues.

If the warning continues to concern you and you'd like a cleaner console output, you could suppress the warning by setting 'torch.use_deterministic_algorithms(False)' earlier in your script, before initializing the training process. Just be aware that this might lead to slight variations if you rerun the training with the same parameters.

Should this warning evolve into an error or if you face any blocking issues during the training, please provide a minimal reproducible example and any relevant information like the output of `torch.__version__` to help us identify the issue more precisely. For now, you can continue with your training as the warning is more of an advisory message and does not indicate that the training will not work.

We appreciate your understanding and your contribution to making our tools better! If anything changes or you encounter other issues, feel free to reach out again.

**Comment by WangFengtu1996** [comment]
I have same error. any update ?

**Comment by glenn-jocher** [comment]
@WangFengtu1996 hey there! 👋 It looks like you're encountering the deterministic algorithms warning while training RT-DETR. This warning is due to certain operations in PyTorch that don't have deterministic implementations when `torch.use_deterministic_algorithms(True)` is set. 

For RT-DETR training, setting `torch.use_deterministic_algorithms(False)` right before your training starts should help avoid this warning, as deterministic behavior isn't strictly necessary unless you require exact reproducibility for debugging or comparison purposes. Here's a quick adjustment to your code:

```python
import torch
from ultralytics import RTDETR

if __name__ == "__main__":
    # It's generally safe to disable deterministic algorithms for training
    torch.use_deterministic_algorithms(False)
    
    model = RTDETR("yolov8-rtdetr.yaml")
    result = model.train(data="VisDrone.yaml", epochs=300, batch=8, imgsz=640, device=0)
```

This change tells PyTorch not to enforce deterministic behavior, which should prevent the warning from appearing. Keep in mind, this might introduce slight variations in training outcomes across different runs. If you have any more questions or run into further issues, feel free to ask. Happy coding! 😊

**Comment by RezaAsadi1991** [comment]
Hello, 
I have this code:

`from ultralytics import YOLO
model = YOLO('yolov8-seg.yaml')
torch.use_deterministic_algorithms(False)
model.train(data='../datasets/Yolov8Format/data.yaml', epochs=500, imgsz=640)`
I tried to integrate a new attention (EMA based) to the network for segmentation. I got this error:

RuntimeError: adaptive_avg_pool2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True)'. You can turn off determinism just for this operation, or you can use the 'warn_only=True' option, if that's acceptable for your application. You can also file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation.

I even utilized the command for deterministic but it did not solve the error and it is not a warning it is an error.
Can you please help me with this issue??

**Comment by mariam-162** [comment]
Hi,

Hello, I'm new to learning and appreciate this community's help. 
I'm using the same code as you mentioned above but the same error is persistent after each epoch. 

 Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size
  0%|          | 0/877 [00:00<?, ?it/s]C:\Users\USER\anaconda3\envs\yolo8help\lib\site-packages\torch\autograd\__init__.py:266: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at C:\cb\pytorch_1000000000000\work\aten\src\ATen\Context.cpp:83.)
  Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
       2/40      13.1G      0.582     0.6146     0.3064         42        640:  37%|███▋      | 321/877 [04:45<08

**Comment by mariam-162** [comment]
Thanks for your prompt response.

I set this flag to false, as you mentioned in earlier comments, and ended up with blue screen (BSOD), PFN_list_corrupt at 27/40 epoch. I'm using Nvidia 3090 with 64G RAM.
The same virtual environment, data set and yaml file ran fine for YOLO v8, 40 epochs.

My code (same as yours):

import torch
from ultralytics import RTDETR

if __name__ == "__main__":
    # It's generally safe to disable deterministic algorithms for training
    torch.use_deterministic_algorithms(False)
    
    model = RTDETR("yolov8-rtdetr.yaml")
    result = model.train(data="VisDrone.yaml", epochs=300, batch=8, imgsz=640, device=0)

Do you know what might have caused the BSOD?

**Comment by glenn-jocher** [comment]
Hey there! 🙌

Sorry to hear about the troubles you've encountered, especially the dreaded blue screen! It's quite rare for training processes to cause a BSOD, but considering it occurred deep into the training (27/40 epoch), this suggests a potential system-level issue, possibly related to hardware drivers or overheating.

Given the complexity of debugging BSOD issues, I'd recommend a few general steps:

- Ensure your Nvidia drivers are up to date. Nvidia frequently releases updates that improve stability and compatibility.
- Monitor your GPU temperature during training. RTDETR and your Nvidia 3090 push the hardware pretty hard. Tools like GPU-Z can help here.
- Test your system stability with other GPU-intensive tasks if possible. This could help identify if the issue is specific to PyTorch/YOLO training or a broader hardware stability issue.

Lastly, sometimes turning off certain BIOS settings like overclocking or adjusting power management settings for the GPU can bring stability.

Let's hope your training runs smoothly next time! Keep us posted.🤞

**Comment by Preference-Kim** [comment]
> Hello there! 👋
> 
> It looks like you're encountering a warning about non-deterministic operations due to `torch.use_deterministic_algorithms(True)`. When training models like YOLOv8 where exact reproducibility isn't crucial, you can safely ignore this warning, or disable deterministic algorithms to prevent it from appearing. Here's how to do it:
> 
> ```python
> import torch
> # Disable deterministic algorithms
> torch.use_deterministic_algorithms(False)
> ```
> 
> Insert this line right after your imports and before initiating the training process. This should help eliminate the warning. Remember, slight variations in training outcomes are normal when deterministic algorithms are disabled. Happy coding and welcome to the community! 😊

Regarding the issue, I was able to resolve it by modifying deterministic to false in training configuration.

Attempts to disable deterministic algorithms by importing torch separately, as mentioned above, did not work for me. This is likely because the Trainer class in Ultralytics changes these settings during initialization. For instance, if you check the constructor of the `BaseTrainer` class, you will notice that it calls the `init_seeds` function, which in turn invokes `torch.use_deterministic_algorithms` with corresponding args.

In conclusion, the solution I found is straightforward: simply set deterministic to False in the config.
I hope this helps :)

**Comment by glenn-jocher** [comment]
Thank you for sharing your solution. Setting deterministic to False in the config is indeed a practical approach. If you have further questions, feel free to ask!

**Comment by Ayazzia01** [comment]
Any updates on resolving this warning?

**Comment by glenn-jocher** [comment]
The warning is related to non-deterministic operations in PyTorch. You can set `deterministic` to False in your training config to address it. If you need further assistance, feel free to ask!

**Comment by darwiinmacas** [comment]
I am concerned about the reproducibility of the results, as I wish to write a paper. what other alternatives would I have or are there not? should I be concerned or are the variations insignificant?

**Comment by glenn-jocher** [comment]
Reproducibility is important for research. While non-deterministic operations can cause slight variations, they are usually minor. For more consistent results, consider setting seeds and using deterministic algorithms where possible.

**Comment by glenn-jocher** [comment]
@darwiinmacas for writing a paper, ensuring reproducibility is important. You can try setting a fixed random seed and using deterministic algorithms where possible, but note that some variations may still occur due to hardware differences. If the variations are minor, they typically won't affect the overall conclusions.

**Comment by glenn-jocher** [comment]
For writing a paper, ensure reproducibility by setting a fixed random seed and using deterministic algorithms where possible. Variations are typically minor, but documenting your setup and any non-deterministic elements is advisable.

**Comment by GoodandNo** [comment]
> > Hello there! 👋
> > It looks like you're encountering a warning about non-deterministic operations due to `torch.use_deterministic_algorithms(True)`. When training models like YOLOv8 where exact reproducibility isn't crucial, you can safely ignore this warning, or disable deterministic algorithms to prevent it from appearing. Here's how to do it:
> > ```python
> > import torch
> > # Disable deterministic algorithms
> > torch.use_deterministic_algorithms(False)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Insert this line right after your imports and before initiating the training process. This should help eliminate the warning. Remember, slight variations in training outcomes are normal when deterministic algorithms are disabled. Happy coding and welcome to the community! 😊
> 
> Regarding the issue, I was able to resolve it by modifying deterministic to false in training configuration.
> 
> Attempts to disable deterministic algorithms by importing torch separately, as mentioned above, did not work for me. This is likely because the Trainer class in Ultralytics changes these settings during initialization. For instance, if you check the constructor of the `BaseTrainer` class, you will notice that it calls the `init_seeds` function, which in turn invokes `torch.use_deterministic_algorithms` with corresponding args.
> 
> In conclusion, the solution I found is straightforward: simply set deterministic to False in the config. I hope this helps :)

@I want to know in which file to modify the setting of this parameter

**Comment by GoodandNo** [comment]
> > Hello there! 👋
> > It looks like you're encountering a warning about non-deterministic operations due to `torch.use_deterministic_algorithms(True)`. When training models like YOLOv8 where exact reproducibility isn't crucial, you can safely ignore this warning, or disable deterministic algorithms to prevent it from appearing. Here's how to do it:
> > ```python
> > import torch
> > # Disable deterministic algorithms
> > torch.use_deterministic_algorithms(False)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Insert this line right after your imports and before initiating the training process. This should help eliminate the warning. Remember, slight variations in training outcomes are normal when deterministic algorithms are disabled. Happy coding and welcome to the community! 😊
> 
> Regarding the issue, I was able to resolve it by modifying deterministic to false in training configuration.
> 
> Attempts to disable deterministic algorithms by importing torch separately, as mentioned above, did not work for me. This is likely because the Trainer class in Ultralytics changes these settings during initialization. For instance, if you check the constructor of the `BaseTrainer` class, you will notice that it calls the `init_seeds` function, which in turn invokes `torch.use_deterministic_algorithms` with corresponding args.
> 
> In conclusion, the solution I found is straightforward: simply set deterministic to False in the config. I hope this helps :)

@Preference-Kim  I want to know in which file to modify the setting of this parameter

**Comment by GoodandNo** [comment]
@Preference-Kim i found it ,thank u

**Comment by GoodandNo** [comment]
i solve the warning,but another problem appeared,how to fix it?anyone who knows it?can't be more appreciate!
![a685c751-0935-4b3c-93a3-5732da27af30](https://github.com/user-attachments/assets/239e5b20-ae51-4433-b7df-b21882b016fa)
![a685c751-0935-4b3c-93a3-5732da27af30](https://github.com/user-attachments/assets/65bdef3f-caac-4b10-9a0c-4cf7edfb2e58)

**Comment by Y-T-G** [comment]
@GoodandNo What's your GPU?

**Comment by GoodandNo** [comment]
GTX 1650 4GB, i'm a fresher in rtdetr..




??
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "ultralytics/ultralytics"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年11月9日(星期六) 凌晨2:46
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [ultralytics/ultralytics] RT-DETR train error (Issue #5978)





 
@GoodandNo What's your GPU?

**Comment by Y-T-G** [comment]
@GoodandNo Set `amp=False`

**Comment by GoodandNo** [comment]
ok,i got it ,i'll try it now,thanks for your help,thanks!!!
&nbsp;


??
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Mohammed ***@***.***&gt;; 
发送时间: 2024年11月9日(星期六) 下午3:23
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [ultralytics/ultralytics] RT-DETR train error (Issue #5978)














 
@GoodandNo Set amp=False

**Comment by GoodandNo** [comment]
ok,fine,there is GPU out of memory,i'll find another GPU to train it,thanks again.best wishs


??
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "ultralytics/ultralytics"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年11月9日(星期六) 下午3:23
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [ultralytics/ultralytics] RT-DETR train error (Issue #5978)





 
@GoodandNo Set amp=False

**Comment by glenn-jocher** [comment]
You're welcome! If you have any more questions or need further assistance, feel free to reach out. Best of luck with your training!
